### PR TITLE
feat(discv5): advertise custom multiaddresses

### LINKED ISSUE #1475: feat: ability to manually specify additional multiaddrs
### Problem
As a node operator, if I have followed the ['listening on websocket to enable connections with waku v2 browser peers ](https://github.com/waku-org/nwaku/blob/019f357dc948da11ec2eae7f6403adfb347e4dce/docs/tutorial/websocket.md) tutorial, my node will not advertise `wss` on disc.

This is a problem in all cases where SSL termination is handled upstream of the `nwaku` / `websockify` node combination, such as:

* Kubernetes deployments using ingress proxies.
* Dappnodes (my case) that use an NGINX HTTPS proxy that abstracts away SSL termination / letsencrypt / DNS config.

### Suggested solution
Provide a configuration option (CLI and environment variable) that allows for manually specifying additional `multiaddrs` that are to be advertised on disc protocols.

### Alternatives considered
It was also considered to use the `--websocket-support:true` flag for normal `ws` (albeit on port `8000`), though in this configuration, using the `light-js` client to attempt to dial the node (being behind a dappnode's HTTPS NGINX proxy, forwarded to the websocket port of `8000`), the `light-js` client failed to dial. It was determined that only with `websockify` forwarding to port `60000` on `nwaku` was I able to dial using `light-js`.

Nonetheless, this alternative would still not solve the issue of correctly advertising `wss` and `ws` on the correct IANA-specified ports.

### Acceptance criteria
For this feature to be deemed accepted, a node operator should be able to:

- [ ] Specify additional `multiaddrs` on which their node is dialable (`ws`, and `wss`).

### PR DESCRIPTION
**Pull request by rymnc**
Should fix #1475, as a follow up to https://github.com/waku-org/nwaku/pull/1509. Makes discv5 broadcast custom-set multiaddrs

### CODE REVIEW

File: waku/v2/node/waku_node.nim
Comment (rymnc) [inline waku/v2/node/waku_node.nim:154]: Hard to constrain these types more, since the upstream lib doesn't expose types like `Discv5TableIpLimit`

File: waku/v2/node/waku_node.nim
Comment (rymnc) [inline waku/v2/node/waku_node.nim:273]: `addBootstrapNodes` needs a mutable variable. The fn can be changed to be without side-effects in a different pr, imo

File: waku/v2/utils/wakuenr.nim
Comment (rymnc) [inline waku/v2/utils/wakuenr.nim:49]: Need the raw field so that we can pass in the appropriate type to discv5 protocol setup. (for additional multiaddresses)

**Comment by LNSD** [comment]
@rymnc I do not agree with moving discv5 to the waku_node module. 

I think that all the discovery mechanisms should be managed and orchestrated by the peer manager. As a consequence, these mechanisms should be:
* Either the `wakunode2` app should register the event handlers to the discovery services and handle them by the peer manager instance.
* Or dependency of the peer manager and passed into the peer manager constructor.

**Comment by rymnc** [comment]
> @rymnc I do not agree with moving discv5 to the waku_node module.
> 
> I think that all the discovery mechanisms should be managed and orchestrated by the peer manager. As a consequence, these mechanisms should be:
> 
> * Either the `wakunode2` app should register the event handlers to the discovery services and handle them by the peer manager instance.
> * Or dependency of the peer manager and passed into the peer manager constructor.

I am not sure if the PeerManager was intended to have this functionality, i.e discovery. It is currently used to manage dialing, adding and removing peers.

Maybe @alrevuelta can weigh in here, was this something you intended for the PeerManager to do?

**Comment by LNSD** [comment]
> I am not sure if the PeerManager was intended to have this functionality, i.e discovery. It is currently used to manage dialing, adding and removing peers.

Note the following from my comment:

> Either the wakunode2 app should register the event handlers to the discovery services and handle them by the peer manager instance.

This would be something like:

```Nim
discoveryMechanism.onPeerDiscovered do (peer: PeerId):
  peerManager.addPeer(peer)
```

File: tests/v2/test_waku_discv5.nim
Comment (alrevuelta) [inline tests/v2/test_waku_discv5.nim:151]: minor suggestion, direction use `generateKey()`
`WakuNode.new(nodeKey=generateKey()))`

(see test_helpers.nim)

File: tests/v2/test_waku_discv5.nim
Comment (alrevuelta) [inline tests/v2/test_waku_discv5.nim:167]: nitpick? do we need to mount the relay? with just dicovery should be fine?

File: waku/v2/node/waku_node.nim
Comment (alrevuelta) [inline waku/v2/node/waku_node.nim:152]: perhaps 7 fields for discv5 is too much? WakuNode.new() is starting to have too many parameters.

regarding discv5TableIpLimit, discv5BucketIpLimit, discv5BitsPerHop not sure if having these configurable is too much. default values should be fine, and whatever is not default should be carefully studied. but out of scope now i guess.


Using `DiscoveryConfig` can perhaps reduce the number of fields there? and if empty defaults to `defaultDiscoveryConfig`
```
  DiscoveryConfig(
    tableIpLimits: TableIpLimits(
      tableIpLimit: tableIpLimit,
      bucketIpLimit: bucketIpLimit),
    bitsPerHop: bitsPerHop
```

File: waku/v2/node/waku_node.nim
Comment (alrevuelta) [inline waku/v2/node/waku_node.nim:292]: related to my other comments can we pass directly the DiscoveryConfig? should simplify things a bit?

**Review by alrevuelta — COMMENTED** [review]
thanks! some comments

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in [185a10a](https://github.com/waku-org/nwaku/pull/1512/commits/185a10ad6637c8c13f2e58c5d7305d290c6a3982), thanks!

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in [185a10a](https://github.com/waku-org/nwaku/pull/1512/commits/185a10ad6637c8c13f2e58c5d7305d290c6a3982), thanks!

File: tests/v2/test_waku_discv5.nim
Comment (rymnc) [reply tests/v2/test_waku_discv5.nim]: Addressed in [185a10a](https://github.com/waku-org/nwaku/pull/1512/commits/185a10ad6637c8c13f2e58c5d7305d290c6a3982), thanks!

File: tests/v2/test_waku_discv5.nim
Comment (rymnc) [reply tests/v2/test_waku_discv5.nim]: Addressed in [185a10a](https://github.com/waku-org/nwaku/pull/1512/commits/185a10ad6637c8c13f2e58c5d7305d290c6a3982), thanks!

File: apps/wakunode2/wakunode2.nim
Comment (jm-clius) [inline apps/wakunode2/wakunode2.nim:292]: May be more readable to assign this to a variable before calling `.new()`

File: waku/v2/node/waku_node.nim
Comment (jm-clius) [inline waku/v2/node/waku_node.nim:154]: Why two arguments here? Perhaps better to have only the `discv5BootstrapEnrs` as a `seq[enr.Record]` and leave it to the application to extract that from config.

File: waku/v2/node/waku_node.nim
Comment (jm-clius) [inline waku/v2/node/waku_node.nim:275]: I think we can perhaps keep the node initialisation a bit cleaner by having this population done by the application? I.e. application provides seq of `enr.Record`.

File: waku/v2/node/waku_node.nim
Comment (jm-clius) [inline waku/v2/node/waku_node.nim:280]: Perhaps change the `WakuDiscoveryV5` constructor to rather take a `seq[MultiAddress]` so the node does not need to know about enr construction?

File: waku/v2/node/waku_node.nim
Comment (jm-clius) [inline waku/v2/node/waku_node.nim:266]: This can be set to the `wakuDiscv5` enr if we have that available (in other words, move the construction to before node construction). We want to consolidate ENRs, so only need to create our own with `initEnr` if `wakuDiscv5` did not provide us with one.

Comment (jm-clius) [inline waku/v2/node/waku_node.nim:266]: Provide constructed `wakuDiscv5` as an argument here.

**Review by jm-clius — COMMENTED** [review]
Thanks, this is more or less the direction I had in mind too. I've added some comments re keeping the config vs initialisation a bit cleaner and consolidating the two different ENRs into one.

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in [74c3622](https://github.com/waku-org/nwaku/pull/1512/commits/74c36223864222f0da14cee19f4a8f9b1f7f3654)

Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in [74c3622](https://github.com/waku-org/nwaku/pull/1512/commits/74c36223864222f0da14cee19f4a8f9b1f7f3654)

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in [74c3622](https://github.com/waku-org/nwaku/pull/1512/commits/74c36223864222f0da14cee19f4a8f9b1f7f3654)

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in [74c3622](https://github.com/waku-org/nwaku/pull/1512/commits/74c36223864222f0da14cee19f4a8f9b1f7f3654)

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: I think it is better set this way, since https://github.com/vacp2p/rfc/pull/566 refers to setting custom shard numbers. In a future PR, we can probably create an enum of `enrFieldPairs` which can be passed into the WakuDiscoveryV5 constructor, which all implement `getRawField` for use with the discv5 protocol.

```nim
type EnrFields = enum
  multiaddrs,
  shards
  
type EnrField* = object
  case key: EnrFields
  of multiaddrs:
    multiaddrs: seq[MultiAddress]
  of shards:
    shards: seq[ShardId]
  
getRawField(enrField: EnrField): seq[byte] =
  # same as getRawField implemented, but for ShardId and MultiAddress both
  if enrField.key == multiaddrs: # or a switch case
    return getRawField(enrField.multiaddrs)
  elif enrField.key == shards
    return getRawField(enrField.shards)

proc new*(T: type WakuDiscoveryV5,
          extIp: Option[ValidIpAddress],
          extTcpPort, extUdpPort: Option[Port],
          bindIP: ValidIpAddress,
          discv5UdpPort: Port,
          bootstrapNodes: seq[string],
          enrAutoUpdate = false,
          privateKey: keys.PrivateKey,
          flags: WakuEnrBitfield,
          enrFields: seq[EnrFields],
          rng: ref HmacDrbgContext,
          discv5Config: protocol.DiscoveryConfig = protocol.defaultDiscoveryConfig): T =
          var enrInitFields = @[(WAKU_ENR_FIELD, @[flags.byte])]
          for enrField in enrFields:
            let rawField = enrField.getRawField()
            enrInitFields.add((enrField.key, rawField))
          ...
```

Since the scaling solution will be put in place soon, I will not change the constructor to accept a seq[Multiaddress], wdyt?

File: apps/wakunode2/wakunode2.nim
Comment (rymnc) [reply apps/wakunode2/wakunode2.nim]: Addressed in [63878f0](https://github.com/waku-org/nwaku/pull/1512/commits/63878f0b5d3d554a36e64368ba8b9c67d495ce8d)

File: waku/v2/node/waku_node.nim
Comment (jm-clius) [reply waku/v2/node/waku_node.nim]: In that case I would either explicitly send the shard index and cluster number to the WakuDiscoveryV5 constructor or provide a field constructor elsewhere. The issue here is that we're mixing where the fields are initialised - `flags` for example is passed in as argument and added the enr fields internally, whereas the `multiaddrs` are not.

Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Okay, I think I understand what you mean now. Addressed in [bffcca7](https://github.com/waku-org/nwaku/pull/1512/commits/bffcca70c8259ef80d2c88813d49074fef3197e6). I do think we should init the fields elsewhere in the future, though.

File: waku/v2/node/waku_node.nim
Comment (LNSD) [reply waku/v2/node/waku_node.nim]: Indeed, this is one of the reasons for not having a discv5 within the waku node. You configure discv5 outside, so you don't have to thread down any configuration.

File: waku/v2/node/waku_node.nim
Comment (LNSD) [inline waku/v2/node/waku_node.nim:15]: Isn't this redundant?

**Review by LNSD — COMMENTED** [review]
As I said in my initial comment, I see a bad idea in moving the discovery mechanisms inside the waku_node module. 

Some comments second this idea: moving the code inside the node module means you have to thread down the configuration. But it also implies that the `WakuNode` type has more than one responsibility (besides of orchestrating the different protocols and configuring the switch it also generates the local ENR, configures the discv5 service and holds the discv5 state). So... you cannot test the `WakuNode` without the discovery instantiated.

In summary, this couples the node code with the discovery code.

----

If everybody agrees, I won't block the PR. But I think we should go in the opposite direction: splitting the content of the big modules instead of merging them into big modules.

**Comment by rymnc** [comment]
Moving this PR to draft to decouple disc and `WakuNode` initialization.

**Comment by jm-clius** [comment]
Agree with @LNSD here. The issue is currently our configuration and initialisation is very convoluted.
- WakuDiscv5 requires some addresses only available in WakuNode initialisation for its own initialisation
- WakuNode requires an initialised WakuDiscv5 to be constructed with a consolidated ENR

Perhaps an increment to make it a bit better is extracting those parts of initialisation to do with constructing addresses and making sure that that's available to both WakuDiscv5 and WakuNode during initialisation? That way WakuDiscv5 initialisation can be decoupled and performed before WakuNode initialisation.

For this, the constructor arguments for WakuDiscv5 may also have to change so that it can be initialised from config.

I know from previously looking at this that there are some intricacies and gotchas, so I'm okay with incrementing here in reasonable steps to something cleaner and decoupled.

**Comment by alrevuelta** [comment]
thanks for addressing the comments, resolved them all.

> Maybe @alrevuelta can weigh in here, was this something you intended for the PeerManager to do?

missed this comment. I think we can keep the peer manager and discovery separated. Peer manager should be agnostic on how the peers are discovered/added to the peer manager.

**Comment by rymnc** [comment]
@LNSD, I've decoupled the discv5 and waku_node setup now. Added a deprecated flag to the original `Wakunode.new()` proc in favor of the `WakuNodeAddrMeta` variant (because there are a lot of usages of `Wakunode.new()`).

**Review by LNSD — CHANGES_REQUESTED** [review]
Please, check my comments :)

File: waku/v2/utils/peers.nim
Comment (rymnc) [reply waku/v2/utils/peers.nim]: Addressed in 8f854af7

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in 70e34bae

File: waku/v2/node/waku_node.nim
Comment (rymnc) [reply waku/v2/node/waku_node.nim]: Addressed in 8f854af7

File: apps/wakunode2/wakunode2.nim
Comment (rymnc) [reply apps/wakunode2/wakunode2.nim]: Addressed in 29a3f116

**Review by LNSD — APPROVED** [review]
LGTM ✅

**Review by jm-clius — APPROVED** [review]
Thanks! I think this is a great step forward to also clean up the confusing address configuration and initialisation. It may be worth adding a unit test suite for other aspects of `NetConfig` (e.g. populating sensible `announcedAddresses`) in future PRs.

**Pull request merged**
